File: easy/94_Binary_Tree_Inorder_Traversal.py
# Given the root of a binary tree, return the inorder traversal of its nodes' values.
#
# Input: root = [1,null,2,3]
# Output: [1,3,2]
#
# Input: root = [1]
# Output: [1]
from typing import Optional, List, Iterator


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root_: Optional[TreeNode]) -> List[int]:
        '''
        Space O(1)
        Time O(n)
        '''

        def inorder(root: TreeNode | None) -> Iterator[TreeNode]:
            stack = [root] if root else []

            while stack:
                node = stack.pop()
                if not node.left and not node.right:
                    yield node
                    continue

                if node.right:
                    stack.append(node.right)
                stack.append(TreeNode(node.val))
                if node.left:
                    stack.append(node.left)

        return [node.val for node in inorder(root_)]

File: go/easy/94_Binary_Tree_Inorder_Traversal.go
package main

import "fmt"

// Given the root of a binary tree, return the inorder traversal of its nodes' values.
//
// Example 1:
// Input: root = [1,null,2,3]
// Output: [1,3,2]
//
// Example 2:
// Input: root = []
// Output: []
//
// Example 3:
// Input: root = [1]
// Output: [1]

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func helper(root *TreeNode, res []int) []int {
	if root == nil {
		return res
	}
	res = helper(root.Left, res)
	res = append(res, root.Val)
	res = helper(root.Right, res)
	return res
}

func inorderTraversal(root *TreeNode) []int {
	return helper(root, []int{})
}

func inorderTraversalStack(root *TreeNode) []int {
	var res []int
	var stack []*TreeNode
	curr := root

	for curr != nil || len(stack) > 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}

		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		res = append(res, curr.Val)

		curr = curr.Right
	}

	return res
}

func main() {
	tree := &TreeNode{
		Val:  1,
		Left: nil,
		Right: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val:   3,
				Left:  nil,
				Right: nil,
			},
			Right: nil,
		},
	}
	fmt.Println(inorderTraversal(tree))
	fmt.Println(inorderTraversalStack(tree))
}
